File: src/test/java/com/github/ccguyka/showupdates/objects/ProjectUpdatesTest.java
package com.github.ccguyka.showupdates.objects;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;

import org.junit.Test;
import com.google.common.collect.Lists;

public class ProjectUpdatesTest {

    @Test
    public void parentUpdateIsSet() {
        DependencyUpdates updates = new DependencyUpdates(
                Lists.newArrayList(
                        mock(ArtifactUpdate.class)));

        ProjectUpdates projectUpdates = ProjectUpdates.builder()
            .withParent(updates)
            .build();

        assertThat(projectUpdates.getParent())
            .isEqualTo(updates);
    }

    @Test
    public void parentUpdateIsOverwritten() {
        DependencyUpdates firstUpdate = new DependencyUpdates(
                Lists.newArrayList(
                        mock(ArtifactUpdate.class, "first update")));
        DependencyUpdates secondUpdate = new DependencyUpdates(
                Lists.newArrayList(
                        mock(ArtifactUpdate.class, "second update")));

        ProjectUpdates projectUpdates = ProjectUpdates.builder()
            .withParent(firstUpdate)
            .withParent(secondUpdate)
            .build();

        assertThat(projectUpdates.getParent())
            .isEqualTo(secondUpdate);
    }

    @Test
    public void dependencyUpdateIsSet() {
        DependencyUpdates updates = new DependencyUpdates(
                Lists.newArrayList(
                        mock(ArtifactUpdate.class),
                        mock(ArtifactUpdate.class)));

        ProjectUpdates projectUpdates = ProjectUpdates.builder()
            .withDependency(updates)
            .build();

        assertThat(projectUpdates.getDependency())
            .isEqualTo(updates);
    }

    @Test
    public void dependencyUpdateIsOverwritten() {
        DependencyUpdates firstUpdate = new DependencyUpdates(
                Lists.newArrayList(
                        mock(ArtifactUpdate.class, "first update A"),
                        mock(ArtifactUpdate.class, "first update B")));
        DependencyUpdates secondUpdate = new DependencyUpdates(
                Lists.newArrayList(
                        mock(ArtifactUpdate.class, "second update A")));

        ProjectUpdates projectUpdates = ProjectUpdates.builder()
            .withDependency(firstUpdate)
            .withDependency(secondUpdate)
            .build();

        assertThat(projectUpdates.getDependency())
            .isEqualTo(secondUpdate);
    }

    @Test
    public void dependencyManagementUpdateIsSet() {
        DependencyUpdates updates = new DependencyUpdates(
                Lists.newArrayList(
                        mock(ArtifactUpdate.class),
                        mock(ArtifactUpdate.class)));

        ProjectUpdates projectUpdates = ProjectUpdates.builder()
            .withDependencyManagement(updates)
            .build();

        assertThat(projectUpdates.getDependencyManagement())
            .isEqualTo(updates);
    }

    @Test
    public void dependencyManagementUpdateIsOverwritten() {
        DependencyUpdates firstUpdate = new DependencyUpdates(
                Lists.newArrayList(
                        mock(ArtifactUpdate.class, "first update A"),
                        mock(ArtifactUpdate.class, "first update B")));
        DependencyUpdates secondUpdate = new DependencyUpdates(
                Lists.newArrayList(
                        mock(ArtifactUpdate.class, "second update A")));

        ProjectUpdates projectUpdates = ProjectUpdates.builder()
            .withDependencyManagement(firstUpdate)
            .withDependencyManagement(secondUpdate)
            .build();

        assertThat(projectUpdates.getDependencyManagement())
            .isEqualTo(secondUpdate);
    }

    @Test
    public void pluginUpdateIsSet() {
        DependencyUpdates updates = new DependencyUpdates(
                Lists.newArrayList(
                        mock(ArtifactUpdate.class),
                        mock(ArtifactUpdate.class)));

        ProjectUpdates projectUpdates = ProjectUpdates.builder()
            .withPlugin(updates)
            .build();

        assertThat(projectUpdates.getPlugin())
            .isEqualTo(updates);
    }

    @Test
    public void pluginUpdateIsOverwritten() {
        DependencyUpdates firstUpdate = new DependencyUpdates(
                Lists.newArrayList(
                        mock(ArtifactUpdate.class, "first update A"),
                        mock(ArtifactUpdate.class, "first update B")));
        DependencyUpdates secondUpdate = new DependencyUpdates(
                Lists.newArrayList(
                        mock(ArtifactUpdate.class, "second update A")));

        ProjectUpdates projectUpdates = ProjectUpdates.builder()
            .withPlugin(firstUpdate)
            .withPlugin(secondUpdate)
            .build();

        assertThat(projectUpdates.getPlugin())
            .isEqualTo(secondUpdate);
    }
}

File: src/test/kotlin/com/github/ccguyka/showupdates/objects/ProjectUpdatesTest.kt
package com.github.ccguyka.showupdates.objects

import com.github.ccguyka.showupdates.objects.ProjectUpdates.Companion.builder
import com.google.common.collect.Lists
import org.assertj.core.api.Assertions
import org.junit.Test
import org.mockito.Mockito

class ProjectUpdatesTest {

    @Test
    fun parentUpdateIsSet() {
        val updates = DependencyUpdates(
                Lists.newArrayList(
                        Mockito.mock(ArtifactUpdate::class.java)))
        val projectUpdates = builder()
                .withParent(updates)
                .build()
        Assertions.assertThat(projectUpdates.parent)
                .isEqualTo(updates)
    }

    @Test
    fun parentUpdateIsOverwritten() {
        val firstUpdate = DependencyUpdates(
                Lists.newArrayList(
                        Mockito.mock(ArtifactUpdate::class.java, "first update")))
        val secondUpdate = DependencyUpdates(
                Lists.newArrayList(
                        Mockito.mock(ArtifactUpdate::class.java, "second update")))
        val projectUpdates = builder()
                .withParent(firstUpdate)
                .withParent(secondUpdate)
                .build()
        Assertions.assertThat(projectUpdates.parent)
                .isEqualTo(secondUpdate)
    }

    @Test
    fun dependencyUpdateIsSet() {
        val updates = DependencyUpdates(
                Lists.newArrayList(
                        Mockito.mock(ArtifactUpdate::class.java),
                        Mockito.mock(ArtifactUpdate::class.java)))
        val projectUpdates = builder()
                .withDependency(updates)
                .build()
        Assertions.assertThat(projectUpdates.dependency)
                .isEqualTo(updates)
    }

    @Test
    fun dependencyUpdateIsOverwritten() {
        val firstUpdate = DependencyUpdates(
                Lists.newArrayList(
                        Mockito.mock(ArtifactUpdate::class.java, "first update A"),
                        Mockito.mock(ArtifactUpdate::class.java, "first update B")))
        val secondUpdate = DependencyUpdates(
                Lists.newArrayList(
                        Mockito.mock(ArtifactUpdate::class.java, "second update A")))
        val projectUpdates = builder()
                .withDependency(firstUpdate)
                .withDependency(secondUpdate)
                .build()
        Assertions.assertThat(projectUpdates.dependency)
                .isEqualTo(secondUpdate)
    }

    @Test
    fun dependencyManagementUpdateIsSet() {
        val updates = DependencyUpdates(
                Lists.newArrayList(
                        Mockito.mock(ArtifactUpdate::class.java),
                        Mockito.mock(ArtifactUpdate::class.java)))
        val projectUpdates = builder()
                .withDependencyManagement(updates)
                .build()
        Assertions.assertThat(projectUpdates.dependencyManagement)
                .isEqualTo(updates)
    }

    @Test
    fun dependencyManagementUpdateIsOverwritten() {
        val firstUpdate = DependencyUpdates(
                Lists.newArrayList(
                        Mockito.mock(ArtifactUpdate::class.java, "first update A"),
                        Mockito.mock(ArtifactUpdate::class.java, "first update B")))
        val secondUpdate = DependencyUpdates(
                Lists.newArrayList(
                        Mockito.mock(ArtifactUpdate::class.java, "second update A")))
        val projectUpdates = builder()
                .withDependencyManagement(firstUpdate)
                .withDependencyManagement(secondUpdate)
                .build()
        Assertions.assertThat(projectUpdates.dependencyManagement)
                .isEqualTo(secondUpdate)
    }

    @Test
    fun pluginUpdateIsSet() {
        val updates = DependencyUpdates(
                Lists.newArrayList(
                        Mockito.mock(ArtifactUpdate::class.java),
                        Mockito.mock(ArtifactUpdate::class.java)))
        val projectUpdates = builder()
                .withPlugin(updates)
                .build()
        Assertions.assertThat(projectUpdates.plugin)
                .isEqualTo(updates)
    }

    @Test
    fun pluginUpdateIsOverwritten() {
        val firstUpdate = DependencyUpdates(
                Lists.newArrayList(
                        Mockito.mock(ArtifactUpdate::class.java, "first update A"),
                        Mockito.mock(ArtifactUpdate::class.java, "first update B")))
        val secondUpdate = DependencyUpdates(
                Lists.newArrayList(
                        Mockito.mock(ArtifactUpdate::class.java, "second update A")))
        val projectUpdates = builder()
                .withPlugin(firstUpdate)
                .withPlugin(secondUpdate)
                .build()
        Assertions.assertThat(projectUpdates.plugin)
                .isEqualTo(secondUpdate)
    }
}
